enterprise/server/remote_execution/container/container.go

```diff
@@ -58,6 +63,43 @@ type Stats struct {
 	PeakMemoryUsageBytes int64
 }
 
+// ContainerMetrics handles Prometheus metrics accounting for CommandContainer
+// instances.
+type ContainerMetrics struct {
+	mu          sync.Mutex
+	memoryUsage map[CommandContainer]int64
+}
+
+func NewContainerMetrics() *ContainerMetrics {
+	return &ContainerMetrics{
+		memoryUsage: make(map[CommandContainer]int64),
+	}
+}
+
+// AddCPUNanos records incremental CPU usage for a container.
+func (m *ContainerMetrics) AddCPUNanos(cpuNanos int64) {
+	const millisPerNanos = 1 / 1e6
+	metrics.RemoteExecutionUsedMilliCPU.Add(float64(cpuNanos) * millisPerNanos)
+}
+
+// SetContainerMemoryUsageBytes sets the current memory usage for the container.
+// Once the container is no longer using memory, the container's memory usage
+// *must* be set to 0 via this method to avoid a memory leak.
+func (m *ContainerMetrics) SetContainerMemoryUsageBytes(c CommandContainer, usageBytes int64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if usageBytes == 0 {
+		delete(m.memoryUsage, c)
+	} else {
+		m.memoryUsage[c] = usageBytes
+	}
+	var total int64
+	for _, v := range m.memoryUsage {
+		total += v
+	}
+	metrics.RemoteExecutionMemoryUsageBytes.Set(float64(total))
```

Comment:
would using a different metric type allow you to avoid this locking? It seems like kind of a bummer to have to lock when recording stats. (even though it's probably not an issue in this case)

Reply:
The other metric types are not really suitable since the intent is for this to reflect the current memory usage, so that we can compare to the current "estimated" memory usage.